src/main/java/build/buildfarm/worker/MatchStage.java

```diff
@@ -48,13 +48,18 @@ private boolean fetch(Operation operation) {
       return false;
     }
     Path execDir = worker.root.resolve(operation.getName());
-    output.offer(new OperationContext(
-        operation,
-        execDir,
-        metadata,
-        action,
-        new ArrayList<>(),
-        new ArrayList<>()));
+    try {
+      output.put(new OperationContext(
+          operation,
+          execDir,
+          metadata,
+          action,
+          new ArrayList<>(),
+          new ArrayList<>()));
+    } catch (InterruptedException intEx) {
+      Thread.currentThread().interrupt();
+      return false;
```

Comment:
But if an interrupt happens here, wouldn't the operation be potentially requeued? https://github.com/bazelbuild/bazel-buildfarm/blob/ca740c29147bfe74cc695730602e2f6607df8fa8/src/main/java/build/buildfarm/instance/stub/StubInstance.java#L339

In general, I'm not sure this interrupt handling preserves the current behavior. Currently, Pipeline.join will catch the exception from any thread, and handle it correctly. But in this change, even though you're calling interrupt(), you're not throwing an exception, and interrupt() on its own will not throw it.
Now, if I understand correctly, the way the code currently works is every single put is follows a successful claim() (except error.put), so it's not supposed to block on put ever anyway. But, just in case the code later changes so that it does -- can you please verify that this will terminate the worker, as it's supposed to?

Reply:
That sounds good, actually. I also don't have better suggestions.